File: frontend/src/lib/api/studio/prds.ts
/**
 * PRDs API
 * Educational Note: Handles AI-generated Product Requirements Documents.
 * PRDs are written incrementally by the agent and stored as markdown files.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';
import { createLogger } from '@/lib/logger';

const log = createLogger('studio-prds-api');

/**
 * PRD job record from the API
 */
export interface PRDJob {
  id: string;
  source_id: string | null;
  source_name: string;
  direction: string;
  status: JobStatus;
  status_message: string;
  error_message: string | null;

  // Plan fields
  document_title: string | null;
  product_name: string | null;
  target_audience: string | null;
  planned_sections: Array<{
    section_id: string;
    title: string;
    description: string;
    priority?: string;
  }>;
  planning_notes: string | null;

  // Progress tracking
  sections_written: number;
  total_sections: number;
  current_section: string | null;

  // Generated content
  markdown_file: string | null;
  markdown_filename: string | null;

  // URLs
  preview_url: string | null;
  download_url: string | null;

  // Metadata
  iterations: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting PRD generation
 */
export interface StartPRDResponse {
  success: boolean;
  job_id?: string;
  message?: string;
  error?: string;
}

/**
 * Response from getting PRD job status
 */
export interface PRDJobStatusResponse {
  success: boolean;
  job?: PRDJob;
  error?: string;
}

/**
 * Response from listing PRD jobs
 */
export interface ListPRDJobsResponse {
  success: boolean;
  jobs: PRDJob[];
  error?: string;
}

/**
 * Response from preview endpoint
 */
export interface PRDPreviewResponse {
  success: boolean;
  document_title?: string;
  product_name?: string;
  sections_written?: number;
  total_sections?: number;
  markdown_content?: string;
  status?: JobStatus;
  error?: string;
}

/**
 * PRDs API
 */
export const prdsAPI = {
  /**
   * Start PRD generation
   */
  async startGeneration(
    projectId: string,
    sourceId: string | null,
    direction?: string,
    parentJobId?: string,
    editInstructions?: string
  ): Promise<StartPRDResponse> {
    try {
      const body: Record<string, string> = {
        direction: direction || 'Create a comprehensive PRD covering all relevant product requirements.',
      };
      if (sourceId) body.source_id = sourceId;
      if (parentJobId) body.parent_job_id = parentJobId;
      if (editInstructions) body.edit_instructions = editInstructions;

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/prd`,
        body
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to start PRD generation');
      throw error;
    }
  },

  /**
   * Get the status of a PRD job
   */
  async getJobStatus(projectId: string, jobId: string): Promise<PRDJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/prd-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get PRD job status');
      throw error;
    }
  },

  /**
   * List all PRD jobs for a project
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListPRDJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/prd-jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to list PRD jobs');
      throw error;
    }
  },

  /**
   * Get PRD preview (markdown content)
   */
  async getPreview(projectId: string, jobId: string): Promise<PRDPreviewResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/prds/${jobId}/preview`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get PRD preview');
      throw error;
    }
  },

  /**
   * Get download URL for PRD
   */
  getDownloadUrl(projectId: string, jobId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/prds/${jobId}/download`;
  },

  /**
   * Poll PRD job status until complete or error
   * Educational Note: Added initial delay to avoid race condition where
   * the job might not be saved to disk yet when polling starts.
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: PRDJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 120 // PRDs can take longer
  ): Promise<PRDJob> {
    let attempts = 0;
    let currentInterval = intervalMs;

    // Initial delay to let the job be created and saved
    await new Promise((resolve) => setTimeout(resolve, 1000));

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      // Handle race condition: job might not exist yet in first few attempts
      if (!response.success || !response.job) {
        if (attempts < 3) {
          // Give the backend more time to create the job
          await new Promise((resolve) => setTimeout(resolve, currentInterval));
          attempts++;
          continue;
        }
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      // Gradually increase interval for long-running jobs
      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('PRD generation timed out');
  },

  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/prds/${jobId}`
      );
      return response.data;
    } catch (error) {
      log.error({ err: error }, 'failed to delete PRD job');
      throw error;
    }
  },
};

File: backend/app/api/studio/prds.py
"""
PRD Generator endpoints - AI-generated Product Requirements Documents.

Educational Note: PRD generation demonstrates the agentic loop pattern:
1. Agent plans the document structure (sections to write)
2. Agent writes sections incrementally using write_prd_section tool
3. Agent signals completion via is_last_section=true flag

Output is Markdown which renders nicely on frontend and can be exported.

Routes:
- POST /projects/<id>/studio/prd                    - Start generation
- GET  /projects/<id>/studio/prd-jobs/<id>          - Job status
- GET  /projects/<id>/studio/prd-jobs               - List jobs
- GET  /projects/<id>/studio/prds/<id>/preview      - Preview markdown content
- GET  /projects/<id>/studio/prds/<id>/download     - Download file (md)
- DELETE /projects/<id>/studio/prds/<id>            - Delete PRD
"""
import io
from flask import jsonify, request, current_app, send_file, Response

from app.api.studio import studio_bp
from app.services.studio_services import studio_index_service
from app.services.integrations.supabase import storage_service
from app.services.auth import require_permission


@studio_bp.route('/projects/<project_id>/studio/prd', methods=['POST'])
@require_permission("studio", "prds")
def generate_prd(project_id: str):
    """
    Start PRD generation (background task).

    Request body:
        {
            "source_id": "source-uuid",
            "direction": "optional user direction/preferences"
        }

    Returns:
        202 Accepted with job_id for polling
    """
    from app.services.ai_agents import prd_agent_service
    from app.services.source_services import source_service
    from app.services.background_services.task_service import task_service
    import uuid

    try:
        data = request.get_json()

        # Source is optional — can generate from direction alone
        source_id = data.get('source_id')
        direction = data.get('direction', '')

        # Get source info (if provided)
        source_name = "Direction Only"
        if source_id:
            source = source_service.get_source(project_id, source_id)
            if not source:
                return jsonify({
                    'success': False,
                    'error': 'Source not found'
                }), 404
            source_name = source.get('name', 'Unknown Source')

        # Edit mode: load parent job's document as context for refinement
        parent_job_id = data.get('parent_job_id')
        edit_instructions = data.get('edit_instructions')
        previous_document = None

        if parent_job_id and not edit_instructions:
            return jsonify({
                'success': False,
                'error': 'edit_instructions is required when parent_job_id is provided'
            }), 400

        if parent_job_id:
            parent_job = studio_index_service.get_prd_job(project_id, parent_job_id)
            if not parent_job or not parent_job.get('markdown_file'):
                return jsonify({
                    'success': False,
                    'error': 'Parent job not found or has no document to edit'
                }), 404

            # Load the full markdown content from storage
            markdown_content = storage_service.download_studio_file(
                project_id, "prds", parent_job_id, parent_job['markdown_file']
            )
            if not markdown_content:
                return jsonify({
                    'success': False,
                    'error': 'Failed to load parent document content for editing'
                }), 500
            previous_document = {
                "document_title": parent_job.get("document_title"),
                "product_name": parent_job.get("product_name"),
                "sections_written": parent_job.get("sections_written", 0),
                "markdown_content": markdown_content
            }

        # Create job
        job_id = str(uuid.uuid4())
        job = studio_index_service.create_prd_job(
            project_id=project_id,
            job_id=job_id,
            source_id=source_id,
            source_name=source_name,
            direction=direction
        )

        # Start background task using centralized task_service
        task_service.submit_task(
            task_type="prd",
            target_id=job_id,
            callable_func=prd_agent_service.prd_agent_service.generate_prd,
            project_id=project_id,
            source_id=source_id,
            job_id=job_id,
            direction=direction,
            previous_document=previous_document,
            edit_instructions=edit_instructions,
        )

        return jsonify({
            'success': True,
            'job_id': job_id,
            'message': 'PRD generation started'
        }), 202  # Accepted

    except Exception as e:
        current_app.logger.error(f"Error starting PRD generation: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start PRD generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/prd-jobs/<job_id>', methods=['GET'])
def get_prd_job_status(project_id: str, job_id: str):
    """
    Get status of a PRD generation job.

    Returns:
        Job object with current status, progress, and results if complete
    """
    try:
        job = studio_index_service.get_prd_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        })

    except Exception as e:
        current_app.logger.error(f"Error getting PRD job status: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to get job status: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/prd-jobs', methods=['GET'])
def list_prd_jobs(project_id: str):
    """
    List all PRD jobs for a project, optionally filtered by source.

    Query params:
        source_id (optional): Filter by source ID

    Returns:
        List of PRD jobs sorted by created_at descending
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_prd_jobs(project_id, source_id)

        return jsonify({
            'success': True,
            'jobs': jobs
        })

    except Exception as e:
        current_app.logger.error(f"Error listing PRD jobs: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to list jobs: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/prds/<job_id>/preview', methods=['GET'])
def preview_prd(project_id: str, job_id: str):
    """
    Preview PRD by returning markdown content.

    Returns:
        JSON with markdown content for rendering on frontend
    """
    try:
        job = studio_index_service.get_prd_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404

        markdown_file = job.get('markdown_file')
        if not markdown_file:
            return jsonify({
                'success': False,
                'error': 'PRD file not yet generated'
            }), 404

        # Read markdown content from Supabase Storage
        markdown_content = storage_service.download_studio_file(
            project_id, "prds", job_id, markdown_file
        )

        if not markdown_content:
            return jsonify({
                'success': False,
                'error': 'PRD file not found'
            }), 404

        return jsonify({
            'success': True,
            'document_title': job.get('document_title', 'Product Requirements Document'),
            'product_name': job.get('product_name'),
            'sections_written': job.get('sections_written', 0),
            'total_sections': job.get('total_sections', 0),
            'markdown_content': markdown_content,
            'status': job.get('status')
        })

    except Exception as e:
        current_app.logger.error(f"Error previewing PRD: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to preview PRD: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/prds/<job_id>/download', methods=['GET'])
def download_prd(project_id: str, job_id: str):
    """
    Download PRD as markdown file.

    Query params:
        format (optional): 'md' (default) - PDF export can be added later

    Returns:
        Markdown file for download
    """
    try:
        job = studio_index_service.get_prd_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404

        markdown_file = job.get('markdown_file')
        if not markdown_file:
            return jsonify({
                'success': False,
                'error': 'PRD file not yet generated'
            }), 404

        # Download from Supabase Storage
        markdown_content = storage_service.download_studio_file(
            project_id, "prds", job_id, markdown_file
        )

        if not markdown_content:
            return jsonify({
                'success': False,
                'error': 'PRD file not found'
            }), 404

        # Create safe filename from document title
        document_title = job.get('document_title', 'PRD')
        safe_title = "".join(c for c in document_title if c.isalnum() or c in " -_").strip()
        if not safe_title:
            safe_title = "PRD"
        download_filename = f"{safe_title}.md"

        # Return as downloadable file
        return send_file(
            io.BytesIO(markdown_content.encode('utf-8')),
            mimetype='text/markdown',
            as_attachment=True,
            download_name=download_filename
        )

    except Exception as e:
        current_app.logger.error(f"Error downloading PRD: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to download PRD: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/prds/<job_id>', methods=['DELETE'])
def delete_prd(project_id: str, job_id: str):
    """
    Delete a PRD and its files.

    Returns:
        Success status
    """
    try:
        # Get job to verify it exists
        job = studio_index_service.get_prd_job(project_id, job_id)
        if not job:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404

        # Delete files from Supabase Storage
        storage_service.delete_studio_job_files(project_id, "prds", job_id)

        # Delete from index
        deleted = studio_index_service.delete_prd_job(project_id, job_id)

        return jsonify({
            'success': deleted,
            'message': 'PRD deleted' if deleted else 'Failed to delete from index'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting PRD: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete PRD: {str(e)}'
        }), 500
